.github/workflows/gen_matrix.pl

```diff
@@ -0,0 +1,191 @@
+%% solver version vector for each defined solver, where again at least
+%% one of each should be designated as the main solver version.
+%%
+%% Adding versions or identifiers to any axis is done with a simple
+%% declarative statement in the first section below.
+%%
+%% The output matrix defines a more efficient selection of tests than
+%% a full cartesian product (because the latter would uselessly test
+%% many redundant configurations).
+%%
+%%  1. Test the main version of each solver on the code build with the
+%%     main version of the compiler, for each operating system
+%%     (solver_main_compiler_main).
+%%
+%%  2. Test the other versions of compiled code with the main version
+%%     of each solver, for each operating system
+%%     (solver_main_compiler_all).
+%%
+%%  3. Test the other versions of each solver on the main compiler
+%%     version of the code, for each operating system
+%%     (solver_all_compiler_main).  Note that the tests run against
+%%     each solver independently, but a test run will test all
+%%     available solvers (and skip any unavailable solvers), so the
+%%     most efficient matrix varies each solver version independently
+%%     (but in parallel) for each test run until all solver versions
+%%     are exhausted.
+%%
+%% This script can be run by hand to see the generated JSON output
+%% using SWI Prolog:
+%%
+%%    $ swipl .github/workflows/gen_matrix.pl
+
+:- initialization(main, main).
+:- use_module(library(http/json)).
+:- use_module(library(clpfd)).
+
+%% Declare every Operating System that should be tested (separately in
+%% sequence)
+os(ubuntu).
+
+%% Declare every Compiler that should be used to build the code
+%% (separately in sequence)
+compiler(ghc).
+
+%% Declare every Solver that should be used (in parallel, not in
+%% sequence: one version of every solver is chosen for each test.
+solver(z3).
+solver(yices).
+solver(stp).
+solver(cvc4).
+solver(boolector).
+solver(abc).
+
+%% Specify the main version of each variable (i.e. the most interesting value of that variable.
+main_version(ubuntu, "ubuntu-latest").
+main_version(ghc, "8.10.7").
+main_version(z3, "4_8_14").
+main_version(yices, "2_6").
+main_version(stp, "2_3_3").
+main_version(cvc4, "1_8").
+main_version(boolector, "3_2_2").
+main_version(abc, "2021_12_30").
+
+%% Specify the versions of the operating system(s), along with the
+%% main version(s) that are of primary interest.
+version(ubuntu, ubuntu_latest).
+
+%% Specify the versions of the compiler(s), along with the main
+%% version that is of primary interest.
+version(ghc, "9.0.2").
+version(ghc, "8.10.7").
+version(ghc, "8.8.4").
+version(ghc, "8.6.5").
+
+version(z3, "4_8_8").
+version(z3, "4_8_9").
+version(z3, "4_8_10").
+version(z3, "4_8_11").
+version(z3, "4_8_12").
+version(z3, "4_8_13").
+version(z3, "4_8_14").
+
+version(yices, "2_6").
+version(yices, "2_6_1").
+
+version(stp, "2_3_3").
+version(stp, "2_3_2").
```

Comment:
`ghc`, `z3`, and `yices`' version numbers are listed in ascending order, but all other tools (such as `stp` here) use descending order. It would be better to pick one convention and use it consistently.

Reply:
Ordering doesn't matter to the code (and I added a comment regarding this) but I do prefer the consistency; thanks for catching this.